Added OUIA documentation (#5511)
patternfly/patternfly-react: packages/react-core/src/helpers/OUIA/OUIA.md

   MUST be expected to be able to be controlled via the same automation.
* An id attribute called `data-ouia-component-id`
  * if there is only one instance of the component on the page at once, it is OPTIONAL
  * if there are multiple instances of a component on the page it MUST be used
  * it MUST be unique within the surrounding context of the component (context may be a
    surrounding component or collection)
  * e.g. A vertical navigation can be expected to only be instantiated once on a page. As such
    it does not need any other identifying factors. Another component type, like a button,
    would be created multiple times on a page and requires some kind of unique identifying id.
  * e.g. An list of items could have tags associated with them. These could be presented in a
    tabular format. Whilst the ids of those tags could be identical, they are also contained
    in rows that make them contextually different. An alternative to this is that the id could
    be prefixed with the item id that it is related to, to create a compound id.
* An attribute called `data-ouia-safe`, which is `True` only when the component is in a static

>   state, i.e. no animations are occurring. At all other times, this value MUST be `False`.

Reviewer: Could you maybe summarize how this section is relevant to Patternfly? Copy/pasting from https://ouia.readthedocs.io is redundant. Removing it is also an option.

[Selected](#list-of-ouia-compliant-components) PatternFly 4 components are `OUIA:Component`
compliant.

### Usage

PatternFly 4 generates `data-ouia-component-type` attribute for selected components automatically
based on the name. For example, for [`Button`](/components/button) `data-ouia-component-type` is
`PF4/Button`. The id of `data-ouia-component-id` attribute is random by default. In order to specify
the concrete id you should add an `ouiaId` prop to the component.

### Example

```js
import React from 'react';
import { Button } from '@patternfly/react-core';

<React.Fragment>
  <Button>Button</Button>

>   

Reviewer: Maybe add a `<br />` between the buttons?

  <Button ouiaId="static_id">Button with ouiaId</Button>
</React.Fragment>
```

## List of OUIA compliant components

### react-core package

> * Alert

Reviewer: Can you make these links like `[Alert](/components/alert)`?

---
id: Open UI Automation
section: developer-resources
---

## About Open UI Automation

Open UI Automation (OUIA) is a specification of certain key guidelines to follow when creating a new
web framework or application. The goal is to ease the burden of creating and maintaining automated
testing environments. OUIA standardizes some HTML attributes and web page
behavior. It helps to create trivial and unambiguous locators of the elements in the DOM for
various automated test suites.

### Why not use `id` or `ARIA`?

`ARIA` is designed as an accessibility specification, and it doesn't account for aspects that are crucial

> for automation such as a component type. `id` attribute might be reserved by some front-end code, and

Reviewer 2: ```suggestion
for automation such as a component type. The `id` attribute might be reserved by some front-end code, and
```

there might be restrictions on changing it. Having a separate id attribute makes it application and
framework independent. Developers and testers have a contract that these attributes are designed for
UI automation.

### Specification parts

The following blocks exist in the specification:

* `OUIA:Component` - components and component frameworks
* `OUIA:Page` - page identification
* `OUIA:PageSafe` - interaction safety

You can find the complete description on the [OUIA website](https://ouia.readthedocs.io).

### Usage

PatternFly 4 generates a `data-ouia-component-type` attribute for selected components automatically
based on the name. For example, the `data-ouia-component-type` attribute for the

> [button component](/components/button) is `PF4/Button`. The id of `data-ouia-component-id`

Reviewer 2: ```suggestion
[button component](/components/button) is `PF4/Button`. The id of the `data-ouia-component-id`
```

---
id: Open UI Automation
section: developer-resources
---

## About Open UI Automation

Open UI Automation (OUIA) is a specification of certain key guidelines to follow when creating a new
web framework or application. The goal is to ease the burden of individuals wishing to create and
maintain automated testing environments. OUIA standardizes some HTML attributes and web page
behavior. It helps to create trivial and unambiguous locators of the elements in the DOM for
various automated test suites.

> ### Why not to use `id`, `ARIA` etc.?

Reviewer 2: ```suggestion
### Why not to use `id`, `ARIA` etc.?
``` 
Can we either replace or remove "etc." ? It's not always easily understood by global audiences, especially for folks whose native language isn't rooted in Latin. 

Also, should "Why not to use" be "Why not use" ?

`ARIA` is designed as an accessibility specification and it doesn't account aspects that are crucial

> for automation such as a component type. `id` attribute might be reserved by some frontend code and

Reviewer 2: ```suggestion
for automation such as a component type. `id` attribute might be reserved by some front-end code and
```

Reviewer 2: Added hyphen to align with how RH documents "front-end" as an adj. in the corporate style guide.

there might be restrictions of changing it. Having a separate id attribute makes it application and
framework independent. Developers and testers have a contract that these attributes are designed for
UI automation.

### Specification parts

The following blocks exist in the specification:

> * `OUIA:Component` - components and component frameworks;

Reviewer 2: ```suggestion
* `OUIA:Component` - components and component frameworks
```

> * `OUIA:Page` - page identification;

Reviewer 2: ```suggestion
* `OUIA:Page` - page identification
```

> * `OUIA:PageSafe` - interaction safety.

Reviewer 2: ```suggestion
* `OUIA:PageSafe` - interaction safety 
``` 

We leave punctuation for bullet lists real simple. We only add periods at the end if they're complete sentences or create complete sentences when combined with the list's intro. In this case, we can get rid of them 👍

> The complete description you can find in [https://ouia.readthedocs.io](https://ouia.readthedocs.io).

Reviewer 2: For accessibility purposes, it's best to use descriptive hyperlinks instead of raw URLs (makes it easier for folks using screen readers). So we change this to something along the lines of: 

You can find the complete description on the [OUIA website](https://ouia.readthedocs.io/en/latest/). 

(not sure if that's how you want to describe this page—feel free to change it for something more accurate!)

[Selected](#list-of-ouia-compliant-components) PatternFly 4 components are `OUIA:Component`
compliant.

### Usage

> PatternFly 4 generates `data-ouia-component-type` attribute for selected components automatically

Reviewer 2: ```suggestion
PatternFly 4 generates a `data-ouia-component-type` attribute for selected components automatically
``` 
Adding the "a" made this sound clearer, but I'm not sure if that changes the technical meaning at all. Feel free to apply as you see fit!

> based on the name. For example, for [`Button`](/components/button) `data-ouia-component-type` is

Reviewer 2: ```suggestion
based on the name. For example, the `data-ouia-component-type` attribute for the [button component]((/components/button) is
```

`PF4/Button`. The id of `data-ouia-component-id` attribute is random by default. To specify the

> concrete id you should add an `ouiaId` prop to the component.

Reviewer 2: ```suggestion
concrete id, you should add an `ouiaId` prop to the component.
```

---
id: Open UI Automation
section: developer-resources
---

## About Open UI Automation

Open UI Automation (OUIA) is a specification of certain key guidelines to follow when creating a new
web framework or application. The goal is to ease the burden of individuals wishing to create and

> maintain automated testing environments. OUIA standardizes some HTML attributes and web page

Reviewer 2: Rewrite: "The goal is to ease the burden of creating and maintaining automated testing environments."

Author: Done.

behavior. It helps to create trivial and unambiguous locators of the elements in the DOM for
various automated test suites.

### Why not use `id` or `ARIA`?

> `ARIA` is designed as an accessibility specification and it doesn't account aspects that are crucial

Reviewer 2: ```suggestion
`ARIA` is designed as an accessibility specification, and it doesn't account for aspects that are crucial
```

> for automation such as a component type. `id` attribute might be reserved by some front-end code and

Reviewer 2: ```suggestion
for automation such as a component type. `id` attribute might be reserved by some front-end code, and
```

> there might be restrictions of changing it. Having a separate id attribute makes it application and

Reviewer 2: ```suggestion
there might be restrictions on changing it. Having a separate id attribute makes it application and
```

framework independent. Developers and testers have a contract that these attributes are designed for
UI automation.

### Specification parts

The following blocks exist in the specification:

* `OUIA:Component` - components and component frameworks
* `OUIA:Page` - page identification
* `OUIA:PageSafe` - interaction safety

You can find the complete description on the [OUIA website](https://ouia.readthedocs.io).

> [Selected](#list-of-ouia-compliant-components) PatternFly 4 components are `OUIA:Component`

Reviewer 2: I feel like we can delete this line and instead just change the list headings that it links to to read "OUIA-compliant PatternFly 4 components"

Author: Fixed.

compliant.

### Usage

PatternFly 4 generates a `data-ouia-component-type` attribute for selected components automatically
based on the name. For example, the `data-ouia-component-type` attribute for the
[button component](/components/button) is `PF4/Button`. The id of `data-ouia-component-id`

> attribute is random by default. To specify the concrete id, you should add an `ouiaId` prop to the

Reviewer 2: ```suggestion
attribute is random by default. To specify the concrete id, add an `ouiaId` prop to the
```

component.

### Example

```js
import React from 'react';
import { Button } from '@patternfly/react-core';

<React.Fragment>
  <Button>Button</Button>
  <br />
  <br />
  <Button ouiaId="static_id">Button with ouiaId</Button>
</React.Fragment>
```

> ## List of OUIA compliant components

Reviewer 2: See above comment on deleting the line about the list and changing this heading to "OUIA-compliant PatternFly 4 components"

Author: Done.
